src/main/java/org/apache/maven/plugins/shade/DefaultShader.java

```diff
@@ -694,6 +702,11 @@ private void addResource( Set<String> resources, JarOutputStream jos, String nam
 
         List<Relocator> relocators;
 
+        // Use thread-local, just in case 'map*' calls are ever done concurrently. Make sure that the using class
+        // initialises this value according to its needs, usually setting the value to false per file before starting
+        // relocation.
+        ThreadLocal<Boolean> wasRelocated = new ThreadLocal<>();
```

Comment:
usually ThreadLocals are declared a "static final" members

Reply:
guess there is no need of any threadlocal, the boolean can be stored in a custom ClassRemapper, sounds simpler/less hacky in terms of code to me, wdyt?

Reply:
@eolivelli, I can make the TL final, no problem.

@rmannibucau, the plugin already uses one `ClassRemapper` per file (i.e. per JAR entry, not per JAR) in order to take care of source file references in each class file. I think we should not create extra objects (one per file in each source JAR) just so as to store a single value. I am not an ASM buff either, BTW. But if the thread-local is not acceptable, I would change it before the PR does not get merged. I am happy to get quick feedback, after my other PR went unnoticed for so long, even though it provides higher value than this one.

Reply:
Update: I forgot to mention, that the custom `Remapper` called `RelocatorRemapper` is instantiated only once, and I did not want to refactor the whole plugin. So I stuck with the idea of one instance and used the thread-local field.

Reply:
A thread local is a hack in the sense it is a solution to keep track of a data in a multithreaded application with reference breakage which is not the case here. Here you have 1 instance of classreader/writer per class file which can track the state of the rewritten class (plain boolean).
An example of such thing can be seen in sirona: https://github.com/rmannibucau/sirona/blob/trunk/agent/javaagent/src/main/java/com/github/rmannibucau/sirona/javaagent/SironaClassVisitor.java#L139
So no additional instantiation in object count, no plugin refactoring or so and simpler code path since the state becomes an attribute and not an unrelated threadlocal so overall I don't see any reason to use a quick and dirty option for this.

hope it makes sense

Reply:
Probably it does, but I do not understand how you would implement it, because the methods which knows that a relocation has taken place is `DefaultShader.RelocatorRemapper.map(String, boolean)`, and that object is only instantiated once like I said. It has no reference to the `ClassRemapper` used locally in `DefaultShader.addRemappedClass()`, but the other way around. If I would pass all the way through to the `map` method, it would not only be ugly but also mean that in the remapper it would have to be kept in a thread-local again, because the (quasi singleton) remapper can be called multiple times.

So if you mean something else and have thought this through, either explain it as clearly as you can, so I can implement it according to your idea, or accept the PR with the thread-local and refactor it by yourself in a later stage. What I am presenting is a substantial improvement. I see no need to make it perfect, just better than before without any new bugs. Let us apply the boyscout rule: leave the camp ground behind a little cleaner than we found it, but let us not stay there longer than the actual trip took.

Reply:
Start by creating a ShadeClassRemapper (a visitor) and each time remapper can be call test it was called or not.
You can even just use a boolean in the remapper  since it is single thread and can be simpler at the end but in terms of arch it is the visitor which would hold that logic by decorating remapper usages.

Another thing to test is just to compare the byte[] before and after, guess in several cases it will be the same and it can be a sufficient first impl but the remapper decoration is what makes the code the more straight forward and less hacky IMHO.

Can surely hack it next week if you need help about this.

Reply:
> Start by creating a ShadeClassRemapper... 

It is after midnight here, maybe I can look into it tomorrow when I can think straight again. Thanks. 

> Another thing to test is just to compare the byte[] before and after

No, that is exactly what is **not** working and what my change is about: to determine that no relocation had taken place, even though the byte code is different due to rewriting with ASM. Those different files, even though I was sure they were unrelated to any relocation, were what triggered my attention and made me discover what was going on here. So that idea is a dead end. Maybe check out and use my code, looking at it as a whole instead of focusing on the thread-local, if you have not done so yet, don't just look at it on GitHub.

Reply:
I did review it, except this threadlocal issue which will explode in a few more iterations at our face it looks acceptable to me

Reply:
I didn't say or mean, you didn't review it. You obviously did. I was not sure if you checked out the branch and watched what it does and compared to what it did before. If you did, fine. Then you know the byte code comparison doesn't cut it. Possibly your other idea does. I never understood this cryptic way of nesting and delegating readers, writers, visitors and remappers in ASM. So this is me being inexperienced in that field or maybe also being not smart enough or not conditioned on this kind of usage pattern. But I do promise to try and understand what you mean tomorrow.

Reply:
Well it does if you rewrite it twice which is likely slower - didnt test but being in mem we can be surprised, was my point.
Agree asm API is not that nice when used in visitor but it is needed copying a class so let's keep it for this iteration maybe.
Can help with code on monday (too late for this week ;))

Reply:
Why would I rewrite it twice, other than in a test? In a test, I could rewrite without relocation, then again with, hoping that that way a comparison would be easier. But we are talking about the production code here.

Reply:
I made the `ThreadLocal` final, as requested. I decided not to refactor to create more extra ASM visitor instances just so as to avoid the thread-local field. I do not see that is should "blow up anyone's face", as @rmannibucau said. The code is simple enough and nothing I am doing there is shady or rocket science. Of course, you can always solve a problem by introducing another layer of complexity, but I have decided against it here and not to cater to someone's personal antipathy for thread-locals. Sorry, but this is my decision for this contribution. We wasted too much time fior discussing this tiny detail already. I would rather have gotten feedback concerning the functionality as such.

@rmannibucau, @eolivelli, would you please also review #88. I have been waiting for a review and for the final merge for too long already. I basically added a new feature for source code shading respecting package exclusions, hence fixing the existing, fragmentary functionality, which was completely broken IMO. Now it is slower, but it does what it should.

Reply:
@kriegaex the thread local issue is that it is easily breakable - it is fragile code, wrongly auto-document the code (threadlocal is mainly about multithreaded code and here you have a single thread) and is easily manipulable by code not belonging to this layer (it is not the remapper which owns that logic but the visitor which actually delegates to the writer or writer itself which creates the class). This is why I say it is a quick and dirty fix/workaround and not a future proof solution. Will work on it on monday forking your branch.

Reply:
Yes, in a way it is premature optimisation. I call it defensive programming. I know that it is single-threaded at the moment, but that could change in the future. The funny thing is, if I would have ignored the thread-unsafety here, nobody would have raised a red flag during the code review. The fact that I did take some precaution, suddenly makes it bad. I disagree with that kind of assessment. But feel free to fork this, if you think you must. I am always happy to learn. I just hink, this should be done in a future PR, not here. I would rather have addressed my question from the TODO comment here, but there the reaction was to do it next time. I guess, different people just have a different sense of priorities, because it is a matter of personal judgement.

I simply disagree with you here about "easily breakable" or "fragile". I am not going to change it.l

Reply:
It will likely not change cause multithreading is unlikely to be faster to write a zip until you load all inputs/output in mem which means risking an oom.
Fact encapsulation is not respected and contract is noy consistent with the usage makes me think it is saner to just impl the feature for whta it is rather than doing a quick fix.
Once again, no issue if _you_ do not change it while it is done ;).

Reply:
> multithreading is unlikely to be faster to write a zip

Good point, I agree. But reading and processing could be done concurrently, writing then queued depending on which thread or pipeline is done with its transformation. I do admit it is a theoretical situation. If after my PR is merged you want to ditch the thread-local in favour of an ordinary boolean (then no longer final in that case, of course) in a subsequent refactoring step, I will not object. The thread-local can always be re-introduced on demand. But at least we have one commit documenting what an (imperfect, but working) solution could look like.

Reply:
Well to have tested such options, unless you have tons of relocations (really tons) it will be slower.
Let's wait a few dayd i PR on your pr with my proposal and move forward from that.

Hope it sounds ok.

Reply:
Actually, after my experience with questions asked on the mailing list, issues opened on Jira and PRs on GitHub, I am kind of afraid that this issue is going to lose momentum and trickle out instead of being merged and closed. Let's not bloat this. We got the code and integration test in place, I would love to see this one merged and then MSHADE-252 next. Still no feedback there, and this is one I really have used for months from my private repository with a temporary version number, because without it Maven Shade is basically useless for me and the AspectJ project, as we also produce shaded and relocated source JARs which should respect excludes, too.